File: src-ts/extractors/TranscriptExtractor.ts
/**
 * Transcript extraction using youtube-transcript-api
 * Extracts transcripts from YouTube videos with rate limiting and retry logic
 */

import { YoutubeTranscript } from 'youtube-transcript';
import logger from '../utils/logger.js';
import { ValidationError, AppError } from '../errors/index.js';

/**
 * Transcript segment with timing
 */
export interface TranscriptSegment {
  start: number;
  duration: number;
  text: string;
}

/**
 * Extracted transcript data
 */
export interface TranscriptData {
  full_text: string;
  segments: TranscriptSegment[];
  language: string;
  is_auto_generated: boolean;
  from_whisper?: boolean;
}

/**
 * Options for transcript extraction
 */
export interface TranscriptExtractorOptions {
  languages?: string[];
  rateLimitDelay?: number;
  whisperExtractor?: any;
}

/**
 * TranscriptExtractor class for extracting transcripts from YouTube
 */
export class TranscriptExtractor {
  private readonly languages: string[];
  private readonly rateLimitDelay: number;
  private lastRequestTime: number;
  private readonly whisperExtractor: any;

  /**
   * Create a new TranscriptExtractor instance
   *
   * @param options - Extractor options
   */
  constructor(options: TranscriptExtractorOptions = {}) {
    this.languages = options.languages || ['en', 'en-GB', 'en-US'];
    this.rateLimitDelay = options.rateLimitDelay || 2000; // 2 seconds default
    this.lastRequestTime = 0;
    this.whisperExtractor = options.whisperExtractor;

    logger.info(
      {
        languages: this.languages,
        rateLimitDelay: this.rateLimitDelay,
        hasWhisperFallback: !!this.whisperExtractor,
      },
      'TranscriptExtractor initialized'
    );
  }

  /**
   * Extract transcript for a video
   *
   * @param videoId - YouTube video ID
   * @param maxRetries - Maximum number of retries on rate limit (default: 3)
   * @param useWhisperFallback - Whether to fallback to Whisper if YouTube fails (default: true)
   * @returns Transcript data or null if unavailable
   * @throws {ValidationError} If video ID is invalid
   * @throws {AppError} If extraction fails after retries
   */
  async extract(
    videoId: string,
    maxRetries = 3,
    useWhisperFallback = true
  ): Promise<TranscriptData | null> {
    // Validate video ID
    if (typeof videoId !== 'string' || !videoId.trim()) {
      throw new ValidationError('videoId must be a non-empty string');
    }

    if (!/^[a-zA-Z0-9_-]{11}$/.test(videoId)) {
      throw new ValidationError('Invalid YouTube video ID format');
    }

    // Rate limiting: wait before making request
    await this.waitForRateLimit();

    // Try YouTube transcript API first
    let retryCount = 0;
    let youtubeFailedReason: string | null = null;

    while (retryCount <= maxRetries) {
      try {
        const transcriptData = await this.extractTranscript(videoId);
        if (transcriptData) {
          logger.info(
            {
              videoId,
              language: transcriptData.language,
              segmentCount: transcriptData.segments.length,
            },
            'Successfully extracted YouTube transcript'
          );
          return transcriptData;
        }
        youtubeFailedReason = 'No transcript available';
        break; // No transcript available, don't retry
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);

        // Check if it's a rate limit error
        if (errorMessage.toLowerCase().includes('rate limit')) {
          retryCount++;
          if (retryCount <= maxRetries) {
            // Exponential backoff: wait longer each time
            const waitTime = this.rateLimitDelay * Math.pow(2, retryCount);
            logger.warn(
              {
                videoId,
                retryCount,
                maxRetries,
                waitTime,
              },
              'Rate limited - applying exponential backoff'
            );
            await this.sleep(waitTime);
          } else {
            youtubeFailedReason = 'Rate limit exceeded';
            logger.error(
              {
                videoId,
                retries: retryCount,
              },
              'YouTube transcript unavailable (rate limit)'
            );
            break;
          }
        } else {
          // Non-rate-limit error, don't retry
          youtubeFailedReason = errorMessage;
          logger.debug(
            {
              videoId,
              error: errorMessage,
            },
            'YouTube transcript extraction failed'
          );
          break;
        }
      }
    }

    // Fallback to Whisper if YouTube transcript failed and Whisper is available
    if (youtubeFailedReason) {
      if (useWhisperFallback && this.whisperExtractor) {
        try {
          logger.info({ videoId }, 'Trying Whisper fallback');
          const whisperTranscript = await this.whisperExtractor.extract(videoId);
          if (whisperTranscript) {
            logger.info({ videoId }, 'Successfully extracted Whisper transcript');
            return whisperTranscript;
          } else {
            logger.warn({ videoId }, 'Whisper transcription failed');
          }
        } catch (error) {
          const errorMessage =
            error instanceof Error ? error.message.split('\n')[0].substring(0, 100) : String(error);
          logger.error(
            {
              videoId,
              error: errorMessage,
            },
            'Whisper extraction error'
          );
        }
      } else {
        logger.info(
          {
            videoId,
            reason: youtubeFailedReason,
          },
          'No transcript available'
        );
      }
    }

    return null;
  }

  /**
   * Extract transcript (internal method without rate limiting logic)
   *
   * @param videoId - YouTube video ID
   * @returns Transcript data or null if unavailable
   */
  private async extractTranscript(videoId: string): Promise<TranscriptData | null> {
    try {
      // Fetch transcript from YouTube
      const transcript = await YoutubeTranscript.fetchTranscript(videoId, {
        lang: this.languages[0], // youtube-transcript takes a single lang
      });

      if (!transcript || transcript.length === 0) {
        return null;
      }

      // Convert to our format
      const segments: TranscriptSegment[] = transcript.map((segment) => ({
        start: segment.offset / 1000, // Convert ms to seconds
        duration: segment.duration / 1000, // Convert ms to seconds
        text: segment.text,
      }));

      // Combine all text
      const fullText = segments.map((seg) => seg.text).join(' ');

      return {
        full_text: fullText,
        segments,
        language: 'en', // youtube-transcript doesn't return language info
        is_auto_generated: true, // youtube-transcript doesn't distinguish
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);

      // Check for specific error types
      if (
        errorMessage.includes('Transcript is disabled') ||
        errorMessage.includes('No transcript')
      ) {
        logger.debug({ videoId }, 'YouTube transcripts disabled for video');
        return null;
      }

      if (errorMessage.includes('Video unavailable')) {
        logger.warn({ videoId }, 'Video unavailable');
        return null;
      }

      // Re-throw for rate limit errors to be handled by retry logic
      if (errorMessage.toLowerCase().includes('rate limit')) {
        throw error;
      }

      // Log other errors but return null (graceful degradation)
      const errorType = error instanceof Error ? error.constructor.name : 'Error';
      logger.debug(
        {
          videoId,
          errorType,
          message: errorMessage.substring(0, 80),
        },
        'Transcript extraction error'
      );

      return null;
    }
  }

  /**
   * Extract transcripts for multiple videos
   *
   * @param videoIds - List of YouTube video IDs
   * @returns Map of video ID to transcript data
   */
  async extractBatch(videoIds: string[]): Promise<Record<string, TranscriptData | null>> {
    if (!Array.isArray(videoIds)) {
      throw new ValidationError('videoIds must be an array');
    }

    const results: Record<string, TranscriptData | null> = {};

    for (const videoId of videoIds) {
      try {
        results[videoId] = await this.extract(videoId);
      } catch (error) {
        logger.error(
          {
            videoId,
            error: error instanceof Error ? error.message : String(error),
          },
          'Failed to extract transcript in batch'
        );
        results[videoId] = null;
      }
    }

    return results;
  }

  /**
   * Wait for rate limit delay
   */
  private async waitForRateLimit(): Promise<void> {
    const currentTime = Date.now();
    const timeSinceLast = currentTime - this.lastRequestTime;

    if (timeSinceLast < this.rateLimitDelay) {
      const waitTime = this.rateLimitDelay - timeSinceLast;
      logger.debug({ waitTime }, 'Rate limit delay');
      await this.sleep(waitTime);
    }

    this.lastRequestTime = Date.now();
  }

  /**
   * Sleep for specified milliseconds
   *
   * @param ms - Milliseconds to sleep
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Format transcript with timestamps
   *
   * @param transcriptData - Transcript data
   * @returns Formatted transcript string
   */
  formatTranscriptWithTimestamps(transcriptData: TranscriptData): string {
    if (!transcriptData || !transcriptData.segments) {
      return '';
    }

    const formattedLines: string[] = [];

    for (const segment of transcriptData.segments) {
      const timestamp = this.formatTimestamp(segment.start);
      const text = segment.text.trim();
      formattedLines.push(`[${timestamp}] ${text}`);
    }

    return formattedLines.join('\n');
  }

  /**
   * Format seconds as MM:SS or HH:MM:SS
   *
   * @param seconds - Time in seconds
   * @returns Formatted timestamp string
   */
  private formatTimestamp(seconds: number): string {
    const totalSeconds = Math.floor(seconds);
    const hours = Math.floor(totalSeconds / 3600);
    const minutes = Math.floor((totalSeconds % 3600) / 60);
    const secs = totalSeconds % 60;

    if (hours > 0) {
      return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
    } else {
      return `${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
    }
  }

  /**
   * Generate YouTube URL with timestamp
   *
   * @param videoId - YouTube video ID
   * @param seconds - Time in seconds
   * @returns YouTube URL with timestamp parameter
   */
  generateYouTubeTimestampUrl(videoId: string, seconds: number): string {
    const timestampSeconds = Math.floor(seconds);
    return `https://youtube.com/watch?v=${videoId}&t=${timestampSeconds}s`;
  }

  /**
   * Calculate statistics about a transcript
   *
   * @param transcriptData - Transcript data
   * @returns Statistics dictionary
   */
  getTranscriptStats(transcriptData: TranscriptData): {
    word_count: number;
    char_count: number;
    segment_count: number;
    duration_seconds: number;
    language: string;
    is_auto_generated: boolean;
  } {
    if (!transcriptData) {
      return {
        word_count: 0,
        char_count: 0,
        segment_count: 0,
        duration_seconds: 0,
        language: 'unknown',
        is_auto_generated: true,
      };
    }

    const fullText = transcriptData.full_text || '';
    const segments = transcriptData.segments || [];

    const wordCount = fullText.split(/\s+/).filter((w) => w).length;
    const charCount = fullText.length;
    const segmentCount = segments.length;

    // Calculate total duration
    let durationSeconds = 0;
    if (segments.length > 0) {
      const lastSegment = segments[segments.length - 1];
      durationSeconds = lastSegment.start + lastSegment.duration;
    }

    return {
      word_count: wordCount,
      char_count: charCount,
      segment_count: segmentCount,
      duration_seconds: Math.floor(durationSeconds),
      language: transcriptData.language || 'unknown',
      is_auto_generated: transcriptData.is_auto_generated ?? true,
    };
  }
}

File: legacy/python/src/extractors/transcript_extractor.py
"""Transcript extraction using youtube-transcript-api"""
import logging
import time
from typing import List, Dict, Optional
from rich.console import Console
from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api._errors import (
    TranscriptsDisabled,
    NoTranscriptFound,
    VideoUnavailable,
    TooManyRequests
)

console = Console()
logger = logging.getLogger(__name__)


class TranscriptExtractor:
    """Extract transcripts from YouTube videos"""
    
    def __init__(self, languages: List[str] = None, rate_limit_delay: float = 2.0, whisper_extractor=None):
        """
        Initialize transcript extractor
        
        Args:
            languages: List of language codes to try (e.g., ['en', 'en-GB', 'en-US'])
            rate_limit_delay: Delay in seconds between requests to avoid rate limiting (default: 2.0)
            whisper_extractor: Optional WhisperTranscriptExtractor for fallback
        """
        self.languages = languages or ['en', 'en-GB', 'en-US']
        self.rate_limit_delay = rate_limit_delay
        self.last_request_time = 0
        self.whisper_extractor = whisper_extractor
    
    def extract(self, video_id: str, max_retries: int = 3, use_whisper_fallback: bool = True) -> Optional[Dict]:
        """
        Extract transcript for a video
        
        Args:
            video_id: YouTube video ID
            max_retries: Maximum number of retries on rate limit (default: 3)
            use_whisper_fallback: Whether to fallback to Whisper if YouTube fails (default: True)
            
        Returns:
            Dictionary with transcript data or None if unavailable
            {
                'full_text': str,
                'segments': List[Dict],
                'language': str,
                'is_auto_generated': bool,
                'from_whisper': bool (optional)
            }
        """
        # Rate limiting: wait before making request
        current_time = time.time()
        time_since_last = current_time - self.last_request_time
        if time_since_last < self.rate_limit_delay:
            time.sleep(self.rate_limit_delay - time_since_last)
        
        self.last_request_time = time.time()
        
        # Try YouTube transcript API first
        retry_count = 0
        youtube_failed_reason = None
        
        while retry_count <= max_retries:
            try:
                transcript_data = self._extract_transcript(video_id)
                if transcript_data:
                    return transcript_data
                youtube_failed_reason = "No transcript available"
                break  # No transcript available, don't retry
            except TooManyRequests:
                retry_count += 1
                if retry_count <= max_retries:
                    # Exponential backoff: wait longer each time
                    wait_time = self.rate_limit_delay * (2 ** retry_count)
                    console.print(f"  [yellow]Rate limited - waiting {wait_time:.1f}s (retry {retry_count}/{max_retries})[/yellow]")
                    time.sleep(wait_time)
                else:
                    youtube_failed_reason = "Rate limit exceeded"
                    console.print(f"  [yellow]YouTube transcript unavailable (rate limit)[/yellow]")
                    break
        
        # Fallback to Whisper if YouTube transcript failed and Whisper is available
        if youtube_failed_reason:
            if use_whisper_fallback and self.whisper_extractor:
                try:
                    console.print(f"  [cyan]Trying Whisper fallback...[/cyan]")
                    whisper_transcript = self.whisper_extractor.extract(video_id)
                    if whisper_transcript:
                        return whisper_transcript
                    else:
                        console.print(f"  [yellow]Whisper transcription failed[/yellow]")
                except Exception as e:
                    error_msg = str(e).split('\n')[0][:100]  # First line only, max 100 chars
                    console.print(f"  [red]Whisper error: {error_msg}[/red]")
            else:
                console.print(f"  [yellow]No transcript available ({youtube_failed_reason})[/yellow]")
        
        return None
    
    def _extract_transcript(self, video_id: str) -> Optional[Dict]:
        """Internal method to extract transcript (without rate limiting logic)"""
        try:
            # Try to get transcript in preferred languages
            transcript_list = YouTubeTranscriptApi.list_transcripts(video_id)
            
            # Try manual transcripts first
            transcript = None
            is_auto_generated = False
            
            try:
                # Try to find a manual transcript in preferred languages
                for lang in self.languages:
                    try:
                        transcript = transcript_list.find_manually_created_transcript([lang])
                        is_auto_generated = False
                        break
                    except NoTranscriptFound:
                        continue
            except (TranscriptsDisabled, NoTranscriptFound, VideoUnavailable):
                # Expected: no manual transcript available; fall through to auto-generated.
                pass
            except Exception:
                logger.exception(
                    "Unexpected error iterating manual transcripts for video_id=%s",
                    video_id,
                )
                # Don't re-raise; let execution continue to auto-generated fallback

            # Fall back to auto-generated transcripts
            if not transcript:
                try:
                    for lang in self.languages:
                        try:
                            transcript = transcript_list.find_generated_transcript([lang])
                            is_auto_generated = True
                            break
                        except NoTranscriptFound:
                            continue
                except (TranscriptsDisabled, NoTranscriptFound, VideoUnavailable):
                    # Expected: no auto-generated transcript available; fall through.
                    pass
                except Exception:
                    logger.exception(
                        "Unexpected error iterating auto-generated transcripts for video_id=%s",
                        video_id,
                    )
                    # Don't re-raise; let execution continue to "any available" fallback
            
            # If still no transcript, try any available transcript
            if not transcript:
                try:
                    # Get first available transcript
                    transcript = next(iter(transcript_list))
                    is_auto_generated = transcript.is_generated
                except StopIteration:
                    return None
            
            # Fetch the actual transcript data
            segments = transcript.fetch()
            
            # Combine all text
            full_text = ' '.join(segment['text'] for segment in segments)
            
            return {
                'full_text': full_text,
                'segments': segments,
                'language': transcript.language_code,
                'is_auto_generated': is_auto_generated
            }
        
        except TranscriptsDisabled:
            console.print(f"  [dim]YouTube transcripts disabled for this video[/dim]")
            return None
        except NoTranscriptFound:
            console.print(f"  [dim]No YouTube transcript found[/dim]")
            return None
        except VideoUnavailable:
            console.print(f"  [red]Video unavailable[/red]")
            return None
        except TooManyRequests:
            # Re-raise to be handled by retry logic
            raise
        except Exception as e:
            # Extract just the first line of error for cleaner output
            error_type = type(e).__name__
            error_msg = str(e).split('\n')[0][:80]  # First line, max 80 chars
            console.print(f"  [dim]Transcript error: {error_type}[/dim]")
            return None
    
    def extract_batch(self, video_ids: List[str]) -> Dict[str, Optional[Dict]]:
        """
        Extract transcripts for multiple videos
        
        Args:
            video_ids: List of YouTube video IDs
            
        Returns:
            Dictionary mapping video_id to transcript data
        """
        results = {}
        
        for video_id in video_ids:
            results[video_id] = self.extract(video_id)
        
        return results
    
    def format_transcript_with_timestamps(self, transcript_data: Dict) -> str:
        """
        Format transcript with clickable timestamps
        
        Args:
            transcript_data: Transcript data dictionary
            
        Returns:
            Formatted transcript string
        """
        if not transcript_data or 'segments' not in transcript_data:
            return ""
        
        formatted_lines = []
        
        for segment in transcript_data['segments']:
            timestamp = self._format_timestamp(segment['start'])
            text = segment['text'].strip()
            formatted_lines.append(f"[{timestamp}] {text}")
        
        return '\n'.join(formatted_lines)
    
    @staticmethod
    def _format_timestamp(seconds: float) -> str:
        """
        Format seconds as MM:SS or HH:MM:SS
        
        Args:
            seconds: Time in seconds
            
        Returns:
            Formatted timestamp string
        """
        seconds = int(seconds)
        hours = seconds // 3600
        minutes = (seconds % 3600) // 60
        secs = seconds % 60
        
        if hours > 0:
            return f"{hours:02d}:{minutes:02d}:{secs:02d}"
        else:
            return f"{minutes:02d}:{secs:02d}"
    
    @staticmethod
    def generate_youtube_timestamp_url(video_id: str, seconds: float) -> str:
        """
        Generate YouTube URL with timestamp
        
        Args:
            video_id: YouTube video ID
            seconds: Time in seconds
            
        Returns:
            YouTube URL with timestamp parameter
        """
        timestamp_seconds = int(seconds)
        return f"https://youtube.com/watch?v={video_id}&t={timestamp_seconds}s"
    
    def get_transcript_stats(self, transcript_data: Dict) -> Dict:
        """
        Calculate statistics about a transcript
        
        Args:
            transcript_data: Transcript data dictionary
            
        Returns:
            Dictionary with statistics
        """
        if not transcript_data:
            return {
                'word_count': 0,
                'char_count': 0,
                'segment_count': 0,
                'duration_seconds': 0
            }
        
        full_text = transcript_data.get('full_text', '')
        segments = transcript_data.get('segments', [])
        
        word_count = len(full_text.split())
        char_count = len(full_text)
        segment_count = len(segments)
        
        # Calculate total duration
        duration_seconds = 0
        if segments:
            last_segment = segments[-1]
            duration_seconds = last_segment['start'] + last_segment.get('duration', 0)
        
        return {
            'word_count': word_count,
            'char_count': char_count,
            'segment_count': segment_count,
            'duration_seconds': int(duration_seconds),
            'language': transcript_data.get('language', 'unknown'),
            'is_auto_generated': transcript_data.get('is_auto_generated', True)
        }


def extract_transcript(video_id: str, languages: List[str] = None) -> Optional[Dict]:
    """
    Convenience function to extract a single transcript
    
    Args:
        video_id: YouTube video ID
        languages: Preferred languages
        
    Returns:
        Transcript data dictionary or None
    """
    extractor = TranscriptExtractor(languages)
    return extractor.extract(video_id)
